### ISSUE TITLE
GitHub credentials aren't saved when prompted from vs code

### ISSUE THREAD

**Issue by xt0rted** [Microsoft/vscode#25952]
- VSCode Version: 1.12
- OS Version: Win10
- Git Version: 2.11.1 then 2.12.2

This was first reported to https://github.com/Microsoft/Git-Credential-Manager-for-Windows/issues/421 but is being cross posted here since it's unclear if the issue is coming from VS Code or the Git Credential Manager.

What's happening is the GCM keeps prompting me to login to GitHub and even when I do the credentials aren't saved to the Windows Credential Manager. I can repeatedly enter them and I'll continue to be prompted to login. Even hitting cancel no longer fully dismisses the prompt because in about 60 seconds it comes back. If I enter my login info to the GCM via the command prompt then they're stored and I'm no longer prompted when opening Code, but if I'm prompted from in Code then they aren't saved.

**Comment by joaomoreno** [comment]
VSCode does not know you already have a credential manager. So, if Git needs credentials, it will assume the position as a credential manager, effectively replacing the GCMforWindows. At the same time, Code isn't in the business of storing credentials, so they never get remembered.

> Once I logged in the credentials showed in the Windows Credential Manager and I'm no longer prompted on the command line or in VS Code.

This is indeed the right fix.

If you don't use the git functionality, as you mentioned in the cross post, simply disable it in the settings: `git.enabled = false`

**Comment by whoisj** [comment]
@joaomoreno this does not seem like quite the correct behavior for VSCode to be doing. Why not have code just send `-c "credential.helper=<path_to_helper>"` as part of the command VSCode sends to Git? This way Git can enumerate all of the credential helpers and work with them as it sees fit.

Another thing to consider here is that Git for Windows ships with GCM preinstalled, and enabled by default. The GCM has multi-factor authentication support for VSTS, GitHub, and Bitbucket - by VSCode attempting to be the authentication utility for Git you're potentially forcing users into less secure authentication models.

**Comment by whoisj** [comment]
Quick update here, I used VSCode to pull from a secure VSTS repository and the output looks promising. Perhaps VSCode does not insert itself as a credential supplier after all.

```
13:59:00.350193 git.c:371               trace: built-in: git 'pull'
13:59:00.353693 run-command.c:369       trace: run_command: 'merge-base' '--fork-point' 'refs/remotes/origin/option' 'option'
13:59:00.408692 git.c:371               trace: built-in: git 'merge-base' '--fork-point' 'refs/remotes/origin/option' 'option'
13:59:00.419692 run-command.c:369       trace: run_command: 'fetch' '--update-head-ok'
13:59:00.474192 git.c:371               trace: built-in: git 'fetch' '--update-head-ok'
13:59:00.476692 run-command.c:369       trace: run_command: 'git-remote-https' 'origin' 'https://dev-x.visualstudio.com/DefaultCollection/_git/Vc2'
13:59:00.852448 run-command.c:369       trace: run_command: 'git credential-manager get'
13:59:00.954439 git.c:596               trace: exec: 'git-credential-manager' 'get'
13:59:00.954439 run-command.c:369       trace: run_command: 'git-credential-manager' 'get'
13:59:00.986941 ...\Program.cs:679      trace: [Main] git-credential-manager (v1.9.1) 'get'
13:59:01.039943 ...\Where.cs:231        trace: [FindGitInstallations] found 1 Git installation(s).
13:59:01.043443 ...Configuration.cs:401 trace: [LoadGitConfiguration] git All config read, 62 entries.
13:59:01.044443 ...\Where.cs:231        trace: [FindGitInstallations] found 1 Git installation(s).
13:59:01.044943 ...Configuration.cs:401 trace: [LoadGitConfiguration] git All config read, 62 entries.
13:59:01.060938 ...\Program.cs:392      trace: [CreateAuthentication] detecting authority type for 'https://dev-x.visualstudio.com/'.
13:59:01.065438 ...uthentication.cs:129 trace: [DetectAuthority] 'https://dev-x.visualstudio.com/' is subdomain of 'visualstudio.com', checking AAD vs MSA.
13:59:01.070938 ...uthentication.cs:230 trace: [GetAuthentication] MSA authority detected.
13:59:01.121463 ...\Program.cs:460      trace: [CreateAuthentication] authority for 'https://dev-x.visualstudio.com/' is Microsoft Live.
13:59:01.128463 ...seSecureStore.cs:134 trace: [ReadCredentials] credentials for 'git:https://dev-x.visualstudio.com' read from store.
13:59:01.128940 ...uthentication.cs:258 trace: [GetCredentials] credentials for 'https://dev-x.visualstudio.com/' found.
13:59:01.141963 ...zureAuthority.cs:172 trace: [ValidateCredentials] validating credentials against 'https://dev-x.visualstudio.com/_apis/connectiondata'.
13:59:01.657582 ...zureAuthority.cs:178 trace: [ValidateCredentials] server returned: 'OK'.
13:59:01.657582 ...\Program.cs:1079     trace: [QueryCredentials] credentials for 'https://dev-x.visualstudio.com/' found.
13:59:01.657582 ...\Program.cs:201      trace: [LogEvent] Microsoft Live credentials for 'https://dev-x.visualstudio.com/' successfully retrieved.
13:59:01.860928 run-command.c:369       trace: run_command: 'git credential-manager store'
13:59:01.962074 git.c:596               trace: exec: 'git-credential-manager' 'store'
13:59:01.963064 run-command.c:369       trace: run_command: 'git-credential-manager' 'store'
13:59:01.995097 ...\Program.cs:679      trace: [Main] git-credential-manager (v1.9.1) 'store'
13:59:02.048573 ...\Where.cs:231        trace: [FindGitInstallations] found 1 Git installation(s).
13:59:02.052574 ...Configuration.cs:401 trace: [LoadGitConfiguration] git All config read, 62 entries.
13:59:02.053573 ...\Where.cs:231        trace: [FindGitInstallations] found 1 Git installation(s).
13:59:02.054073 ...Configuration.cs:401 trace: [LoadGitConfiguration] git All config read, 62 entries.
13:59:02.066101 ...\Program.cs:392      trace: [CreateAuthentication] detecting authority type for 'https://dev-x.visualstudio.com/'.
13:59:02.071100 ...uthentication.cs:129 trace: [DetectAuthority] 'https://dev-x.visualstudio.com/' is subdomain of 'visualstudio.com', checking AAD vs MSA.
13:59:02.077101 ...uthentication.cs:230 trace: [GetAuthentication] MSA authority detected.
13:59:02.126075 ...\Program.cs:460      trace: [CreateAuthentication] authority for 'https://dev-x.visualstudio.com/' is Microsoft Live.
13:59:02.127076 ...\Program.cs:391      trace: [Store] storing VSTS credentials for 'https://dev-x.visualstudio.com/'.
13:59:02.150580 run-command.c:369       trace: run_command: 'rev-list' '--objects' '--stdin' '--not' '--all' '--quiet'
13:59:02.213085 git.c:371               trace: built-in: git 'rev-list' '--objects' '--stdin' '--not' '--all' '--quiet'
13:59:02.225087 run-command.c:369       trace: run_command: 'rev-list' '--objects' '--stdin' '--not' '--all' '--quiet'
13:59:02.287101 git.c:371               trace: built-in: git 'rev-list' '--objects' '--stdin' '--not' '--all' '--quiet'
13:59:02.303100 run-command.c:1155      run_processes_parallel: preparing to run up to 1 tasks
13:59:02.303100 run-command.c:1187      run_processes_parallel: done
13:59:02.303100 run-command.c:369       trace: run_command: 'gc' '--auto'
13:59:02.360116 git.c:371               trace: built-in: git 'gc' '--auto'
13:59:02.372114 run-command.c:369       trace: run_command: 'merge' '--ff-only' 'FETCH_HEAD'
13:59:02.428113 git.c:371               trace: built-in: git 'merge' '--ff-only' 'FETCH_HEAD'
13:59:02.534125 git.c:371               trace: built-in: git 'status' '-z' '-u'
13:59:02.635138 git.c:371               trace: built-in: git 'symbolic-ref' '--short' 'HEAD'
13:59:02.733041 git.c:371               trace: built-in: git 'rev-parse' 'option'
13:59:02.831611 git.c:371               trace: built-in: git 'rev-parse' '--symbolic-full-name' '--abbrev-ref' 'option@{u}'
13:59:02.932197 git.c:371               trace: built-in: git 'rev-list' '--left-right' 'option...origin/option'
13:59:03.032205 git.c:371               trace: built-in: git 'for-each-ref' '--format' '%(refname) %(objectname)'
13:59:03.048708 git.c:371               trace: built-in: git 'remote' '--verbose'
```

Also tested a non-public GitHub repository.

```
14:01:14.486173 git.c:371               trace: built-in: git 'pull'
14:01:14.538024 run-command.c:369       trace: run_command: 'merge-base' '--fork-point' 'refs/remotes/src/master' 'master'
14:01:14.599026 git.c:371               trace: built-in: git 'merge-base' '--fork-point' 'refs/remotes/src/master' 'master'
14:01:14.610030 run-command.c:369       trace: run_command: 'fetch' '--update-head-ok'
14:01:14.670032 git.c:371               trace: built-in: git 'fetch' '--update-head-ok'
14:01:14.673032 run-command.c:369       trace: run_command: 'git-remote-https' 'src' 'https://github.com/Microsoft/Git-Process-Management.git'
14:01:15.235187 run-command.c:369       trace: run_command: 'git credential-manager get'
14:01:15.356784 git.c:596               trace: exec: 'git-credential-manager' 'get'
14:01:15.356784 run-command.c:369       trace: run_command: 'git-credential-manager' 'get'
14:01:15.396786 ...\Program.cs:679      trace: [Main] git-credential-manager (v1.9.1) 'get'
14:01:15.451820 ...\Where.cs:231        trace: [FindGitInstallations] found 1 Git installation(s).
14:01:15.455819 ...Configuration.cs:401 trace: [LoadGitConfiguration] git All config read, 70 entries.
14:01:15.456819 ...\Where.cs:231        trace: [FindGitInstallations] found 1 Git installation(s).
14:01:15.456819 ...Configuration.cs:401 trace: [LoadGitConfiguration] git All config read, 70 entries.
14:01:15.472842 ...\Program.cs:392      trace: [CreateAuthentication] detecting authority type for 'https://github.com/'.
14:01:15.480819 ...uthentication.cs:126 trace: [GetAuthentication] created GitHub authentication for 'https://github.com/'.
14:01:15.480819 ...\Program.cs:450      trace: [CreateAuthentication] authority for 'https://github.com/' is GitHub.
14:01:15.500821 ...seSecureStore.cs:134 trace: [ReadCredentials] credentials for 'git:https://github.com' read from store.
14:01:15.500821 ...uthentication.cs:153 trace: [GetCredentials] credentials for 'https://github.com/' found.
14:01:16.152902 ...\Authority.cs:206    trace: [ValidateCredentials] credential validation for 'https://github.com/' succeeded.
14:01:16.153901 ...\Program.cs:1109     trace: [QueryCredentials] credentials for 'https://github.com/' found.
14:01:16.153901 ...\Program.cs:201      trace: [LogEvent] GitHub credentials for 'https://github.com/' successfully retrieved.
14:01:16.399659 run-command.c:369       trace: run_command: 'git credential-manager store'
14:01:16.508751 git.c:596               trace: exec: 'git-credential-manager' 'store'
14:01:16.509250 run-command.c:369       trace: run_command: 'git-credential-manager' 'store'
14:01:16.545812 ...\Program.cs:679      trace: [Main] git-credential-manager (v1.9.1) 'store'
14:01:16.598836 ...\Where.cs:231        trace: [FindGitInstallations] found 1 Git installation(s).
14:01:16.602814 ...Configuration.cs:401 trace: [LoadGitConfiguration] git All config read, 70 entries.
14:01:16.603814 ...\Where.cs:231        trace: [FindGitInstallations] found 1 Git installation(s).
14:01:16.604814 ...Configuration.cs:401 trace: [LoadGitConfiguration] git All config read, 70 entries.
14:01:16.617321 ...\Program.cs:392      trace: [CreateAuthentication] detecting authority type for 'https://github.com/'.
14:01:16.625325 ...uthentication.cs:126 trace: [GetAuthentication] created GitHub authentication for 'https://github.com/'.
14:01:16.625325 ...\Program.cs:450      trace: [CreateAuthentication] authority for 'https://github.com/' is GitHub.
14:01:16.625325 ...\Program.cs:395      trace: [Store] storing GitHub credentials for 'https://github.com/'.
14:01:16.634324 ...seSecureStore.cs:210 trace: [WriteCredential] credentials for 'git:https://github.com' written to store.
14:01:16.651325 run-command.c:369       trace: run_command: 'rev-list' '--objects' '--stdin' '--not' '--all' '--quiet'
14:01:16.714331 git.c:371               trace: built-in: git 'rev-list' '--objects' '--stdin' '--not' '--all' '--quiet'
14:01:16.725336 run-command.c:369       trace: run_command: 'rev-list' '--objects' '--stdin' '--not' '--all' '--quiet'
14:01:16.799335 git.c:371               trace: built-in: git 'rev-list' '--objects' '--stdin' '--not' '--all' '--quiet'
14:01:16.814824 run-command.c:1155      run_processes_parallel: preparing to run up to 1 tasks
14:01:16.814824 run-command.c:1187      run_processes_parallel: done
14:01:16.814824 run-command.c:369       trace: run_command: 'gc' '--auto'
14:01:16.874384 git.c:371               trace: built-in: git 'gc' '--auto'
14:01:16.886347 run-command.c:369       trace: run_command: 'merge' '--ff-only' 'FETCH_HEAD'
14:01:16.946410 git.c:371               trace: built-in: git 'merge' '--ff-only' 'FETCH_HEAD'
14:01:17.067592 git.c:371               trace: built-in: git 'status' '-z' '-u'
14:01:17.205134 git.c:371               trace: built-in: git 'symbolic-ref' '--short' 'HEAD'
14:01:17.315167 git.c:371               trace: built-in: git 'rev-parse' 'master'
14:01:17.424609 git.c:371               trace: built-in: git 'rev-parse' '--symbolic-full-name' '--abbrev-ref' 'master@{u}'
14:01:17.560614 git.c:371               trace: built-in: git 'rev-list' '--left-right' 'master...src/master'
14:01:17.674146 git.c:371               trace: built-in: git 'for-each-ref' '--format' '%(refname) %(objectname)'
14:01:17.693308 git.c:371               trace: built-in: git 'remote' '--verbose'
```

These results run completely counter to the statement "So, if Git needs credentials, it will assume the position as a credential manager, effectively replacing the GCMforWindows" from @joaomoreno . Hopefully he can clarify.

**Comment by xt0rted** [comment]
I see two different things going on here.

The first being the change in behavior after installing 1.12. Since then the GCM continues to show up whether I enter my credentials or not every 60 seconds or so. The prompt simply won't go away.

The second is how even after entering my credentials into the same prompt I'd get from the console the credentials aren't persisted. The user experience stinks because you enter them thinking they're being saved, since that's the point of the credential manager and VSCode using the system's version of Git, but they aren't which leads me back to the first issue...

The only reason I never bothered with the built-in Git features is because the prompt never stored my credentials and would always show on first run. So I gave up and continued to use GitHub Desktop. I have 2FA enabled so the prompt is useful.

**Comment by whoisj** [comment]
Looks like GfW v2.12.2 shipped with GVM v1.9.0 and v1.9.1 contained a fix for credentials not being persisted for some users. @xt0rted any chance you could download and run the v1.9.1 or v1.10.0.preview.2 installer and see if your issue gets sorted?

Thanks!

**Comment by xt0rted** [comment]
@whoisj 1.9.1 worked.

I cleared my GitHub credentials and this time they were persisted after entering them from VSCode.

If I clear the credentials and select Cancel on the prompt I'm still prompted a minute or so later to login and that goes on until I close out of VSCode.

The constant prompting started in 1.12 even before I updated my version of Git which is why I still think there's an issue in this version, but persisting the credentials is now fixed with 1.9.1 of GCM.

**Comment by joaomoreno** [comment]
> Perhaps VSCode does not insert itself as a credential supplier after all.

You're right, I said it wrong when mentioning that _ it will assume the position as a credential manager_. It only takes the role of `ASKPASS`, which is the last resort in the whole process to ask the user for credentials. Sorry for misleading you there.
